BUG: OmniSciDB Millisecond extraction doesn't work

OmniSciDB Millisecond doesn't work, probably because it is using the Impala translation.

```
E           Exception: Exception: Parse failed: Encountered "\"timestamp_col\"" at line 1, column 16.
E           Was expecting one of:
E               "CENTURY" ...
E               "DAY" ...
E               "DECADE" ...
E               "DOW" ...
E               "DOY" ...
E               "EPOCH" ...
E               "HOUR" ...
E               "ISODOW" ...
E               "ISOYEAR" ...
E               "MICROSECOND" ...
E               "MILLISECOND" ...
E               "MILLENNIUM" ...
E               "MINUTE" ...
E               "MONTH" ...
E               "NANOSECOND" ...
E               "QUARTER" ...
E               "SECOND" ...
E               "WEEK" ...
E               "YEAR" ...
E               : SELECT extract("timestamp_col", 'millisecond') AS tmp
E           FROM functional_alltypes

ibis/omniscidb/client.py:845: Exception
```

--- a/docs/source/release.rst
+++ b/docs/source/release.rst
@@ -13,6 +13,7 @@ Release Notes
 * :feature:`2126` Add translation rules for isnull() and notnull() for pyspark backend
 * :feature:`2062` Implement read_csv for omniscidb backend
 * :feature:`2097` Date, DateDiff and TimestampDiff implementations for OmniSciDB
+* :bug:`2170` Fix millisecond issue for OmniSciDB #2167, MySQL #2169, PostgreSQL #2166 and Pandas #2168 backends
 * :feature:`2060` Add initial support for ibis.random function
 * :support:`2107` Added fragment_size to table creation for OmniSciDB
 * :feature:`2117` Add non-nullable info to schema output

--- a/docs/source/release.rst
+++ b/docs/source/release.rst
@@ -13,7 +13,7 @@ Release Notes
 * :feature:`2126` Add translation rules for isnull() and notnull() for pyspark backend
 * :feature:`2062` Implement read_csv for omniscidb backend
 * :feature:`2097` Date, DateDiff and TimestampDiff implementations for OmniSciDB
-* :bug:`2170` Fix millisecond issue for OmniSciDB #2167, MySQL #2169, PostgreSQL #2166 and Pandas #2168 backends
+* :bug:`2170` Fix millisecond issue for OmniSciDB :issue:`2167`, MySQL :issue:`2169`, PostgreSQL :issue:`2166` and Pandas :issue:`2168` backends
 * :feature:`2060` Add initial support for ibis.random function
 * :support:`2107` Added fragment_size to table creation for OmniSciDB
 * :feature:`2117` Add non-nullable info to schema output

--- a/ibis/omniscidb/operations.py
+++ b/ibis/omniscidb/operations.py
@@ -265,10 +265,23 @@ def _call(translator, func, *args):
 
 
 def _extract_field(sql_attr):
-    def extract_field_formatter(translator, expr):
+    def extract_field_formatter(translator, expr, sql_attr=sql_attr):
+        adjustments = {
+            'MILLISECOND': 1000,
+        }
+        adjustment = adjustments.get(sql_attr, None)
+
         op = expr.op()
-        arg = translator.translate(op.args[0])
-        return 'EXTRACT({} FROM {})'.format(sql_attr, arg)
+        arg = op.args[0]
+
+        arg_str = translator.translate(arg)
+        result = 'EXTRACT({} FROM {})'.format(sql_attr, arg_str)
+
+        if adjustment:
+            # used by time extraction
+            result = ' mod({}, {})'.format(result, adjustment)
+
+        return result
 
     return extract_field_formatter
 
@@ -993,6 +1006,7 @@ _date_ops = {
     ops.ExtractHour: _extract_field('HOUR'),
     ops.ExtractMinute: _extract_field('MINUTE'),
     ops.ExtractSecond: _extract_field('SECOND'),
+    ops.ExtractMillisecond: _extract_field('MILLISECOND'),
     ops.IntervalAdd: _interval_from_integer,
     ops.IntervalFromInteger: _interval_from_integer,
     ops.DateAdd: _timestamp_op('TIMESTAMPADD'),

--- a/ibis/pandas/execution/temporal.py
+++ b/ibis/pandas/execution/temporal.py
@@ -41,6 +41,8 @@ def execute_extract_millisecond_timestamp(op, data, **kwargs):
 @execute_node.register(ops.ExtractTemporalField, pd.Series)
 def execute_extract_timestamp_field_series(op, data, **kwargs):
     field_name = type(op).__name__.lower().replace('extract', '')
+    if field_name == 'millisecond':
+        return (data.dt.microsecond // 1000).astype(np.int32)
     return getattr(data.dt, field_name).astype(np.int32)
 
 

--- a/ibis/pandas/execution/temporal.py
+++ b/ibis/pandas/execution/temporal.py
@@ -33,17 +33,20 @@ def execute_extract_timestamp_field_timestamp(op, data, **kwargs):
     return getattr(data, field_name)
 
 
+@execute_node.register(ops.ExtractTemporalField, pd.Series)
+def execute_extract_timestamp_field_series(op, data, **kwargs):
+    field_name = type(op).__name__.lower().replace('extract', '')
+    return getattr(data.dt, field_name).astype(np.int32)
+
+
 @execute_node.register(ops.ExtractMillisecond, pd.Timestamp)
 def execute_extract_millisecond_timestamp(op, data, **kwargs):
     return int(data.microsecond // 1000.0)
 
 
-@execute_node.register(ops.ExtractTemporalField, pd.Series)
-def execute_extract_timestamp_field_series(op, data, **kwargs):
-    field_name = type(op).__name__.lower().replace('extract', '')
-    if field_name == 'millisecond':
-        return (data.dt.microsecond // 1000).astype(np.int32)
-    return getattr(data.dt, field_name).astype(np.int32)
+@execute_node.register(ops.ExtractMillisecond, pd.Series)
+def execute_extract_millisecond_series(op, data, **kwargs):
+    return (data.dt.microsecond // 1000).astype(np.int32)
 
 
 @execute_node.register(

--- a/ibis/sql/mysql/compiler.py
+++ b/ibis/sql/mysql/compiler.py
@@ -55,6 +55,8 @@ def _extract(fmt):
     def translator(t, expr):
         (arg,) = expr.op().args
         sa_arg = t.translate(arg)
+        if fmt == 'millisecond':
+            return sa.extract('microsecond', sa_arg) % 1000
         return sa.extract(fmt, sa_arg)
 
     return translator

--- a/ibis/sql/postgres/compiler.py
+++ b/ibis/sql/postgres/compiler.py
@@ -87,7 +87,10 @@ def _millisecond(t, expr):
     # we get total number of milliseconds including seconds with extract so we
     # mod 1000
     (sa_arg,) = map(t.translate, expr.op().args)
-    return sa.cast(sa.extract('millisecond', sa_arg), sa.SMALLINT) % 1000
+    return (
+        sa.cast(sa.func.floor(sa.extract('millisecond', sa_arg)), sa.SMALLINT)
+        % 1000
+    )
 
 
 _truncate_precisions = {

--- a/ibis/tests/all/test_temporal.py
+++ b/ibis/tests/all/test_temporal.py
@@ -36,12 +36,20 @@ def test_date_extract(backend, alltypes, df, attr):
 
 
 @pytest.mark.parametrize(
-    'attr', ['year', 'month', 'day', 'hour', 'minute', 'second']
+    'attr', ['year', 'month', 'day', 'hour', 'minute', 'second', 'millisecond']
 )
 @pytest.mark.xfail_unsupported
 def test_timestamp_extract(backend, alltypes, df, attr):
+    if attr == 'millisecond':
+        if backend.name == 'sqlite':
+            pytest.xfail(reason=('Issue #2156'))
+        if backend.name == 'spark':
+            pytest.xfail(reason='Issue #2159')
+        expected = (df.timestamp_col.dt.microsecond // 1000).astype('int32')
+    else:
+        expected = getattr(df.timestamp_col.dt, attr).astype('int32')
+
     expr = getattr(alltypes.timestamp_col, attr)()
-    expected = getattr(df.timestamp_col.dt, attr).astype('int32')
 
     result = expr.execute()
     expected = backend.default_series_rename(expected)

